Rename Normalize and move to nn

jla524 opened the pull request:
![rendered](https://user-images.githubusercontent.com/39754370/200094170-cba5b713-ad12-4816-bf33-394fb1f44f3e.png)

geohot commented:
Great! Add some tests to confirm it matches PyTorch GroupNorm and I'll merge

jla524 commented:
Thanks, I've added a test case for it (very similar to `test_conv2d`). Interestingly, the parameters are reversed in PyTorch, so it looks something like this:
```
class GroupNorm:
    def __init__(self, num_groups: int, num_channels: int):
        ...
```

sachin9889 commented:
class GroupNorm:
    def __init__(self, num_groups: int, num_channels: int):
        ...

jla524 commented:
> Want to flip it to match PyTorch?

Honestly, I think our implementation is more intuitive. Take a look at [this example](https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html):
```
>>> input = torch.randn(20, 6, 10, 10)
>>> # Separate 6 channels into 3 groups
>>> m = nn.GroupNorm(3, 6)
>>> # Separate 6 channels into 6 groups (equivalent with InstanceNorm)
>>> m = nn.GroupNorm(6, 6)
>>> # Put all 6 channels into a single group (equivalent with LayerNorm)
>>> m = nn.GroupNorm(1, 6)
```
It's kinda confusing to pass in the group, and then channel. What do you think of this?

vaibhav4516 reviewed (commented):
ok

geohot commented:
I really don't want it different from PyTorch, I agree ours matches the comments, but people shouldn't have to read the docs if they are migrating, it should just do what they expect.

jla524 commented:
Understood. I'll flip the parameters to match PyTorch.

geohot commented:
"elementwise_affine on channels" HACK needs to be fixed too, should match the PyTorch behavior in all cases.

jla524 commented:
Updated the parameters. I removed the default value for `num_groups` to match PyTorch as well, but had to hard code the blocks in stable diffusion.

The test unfortunately failed, and I can't reproduce it on my device right now :(

jla524 commented:
> "elementwise_affine on channels" HACK needs to be fixed too, should match the PyTorch behavior in all cases.

Sounds good, I'll try and fix it

geohot commented:
Please match exactly the first 4 args of the PyTorch version.

https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html

jla524 commented:
Hi George, thanks for the shout out on stream! Looks like PyTorch doesn't support `num_groups=None`. It must be an integer divisible by `num_channels`. Below is the error message I got when I tried to run it:
```
================================================ FAILURES ================================================
_________________________________________ TestNN.test_groupnorm __________________________________________

self = <test.test_nn.TestNN testMethod=test_groupnorm>

    def test_groupnorm(self):
      # (1, 256, 1280) with G=None C=1280
      BS, H, W, C, G = 1, 16, 16, 320, None
    
      # create in tinygrad
      layer = GroupNorm(G, C)
    
      # create in torch
      with torch.no_grad():
>       torch_layer = torch.nn.GroupNorm(G, C).eval()

test/test_nn.py:106: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[KeyError('num_groups') raised in repr()] GroupNorm object at 0x1104ba980>, num_groups = None
num_channels = 320, eps = 1e-05, affine = True, device = None, dtype = None

    def __init__(self, num_groups: int, num_channels: int, eps: float = 1e-5, affine: bool = True,
                 device=None, dtype=None) -> None:
        factory_kwargs = {'device': device, 'dtype': dtype}
        super(GroupNorm, self).__init__()
>       if num_channels % num_groups != 0:
E       TypeError: unsupported operand type(s) for %: 'int' and 'NoneType'

/usr/local/lib/python3.10/site-packages/torch/nn/modules/normalization.py:250: TypeError
```
We can match this behavior, but it will break stable diffusion. How should I proceed with this?

geohot commented:
Fix stable diffusion in a way PyTorch would like! What is None supposed to do it's been a while since I wrote this

jla524 commented:
Sounds good. I think I can try and reshape the inputs in stable diffusion.

Usually when we call GroupNorm, we expect x to have shape [N, C, H, W]. We reshape x to [N, num_groups, -1], normalize, and reshape back to [N, C, H, W].

But when `num_group=None`, we expect x to have shape [N, num_groups, -1] already, so we only normalize and don't need to reshape.

geohot commented on tinygrad/nn/__init__.py:
```diff
@@ -59,3 +59,22 @@ def __init__(self, in_features, out_features, bias=True):
+    # elementwise_affine on channels
+    if len(x.shape) == 4:
+      # HACK for channels in conv
+      return (x * self.weight.reshape(1, -1, 1, 1)) + self.bias.reshape(1, -1, 1, 1)
```
If self.weight is None here, this will crash.

geohot commented:
Congrats!

geohot commented:
```
kafka@tubby:~/fun/tinygrad$ IMAGE=2 GPU=1 python3 examples/stable_diffusion.py 
issue assigning object on 699
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1145/1145 [00:01<00:00, 722.05it/s]
Traceback (most recent call last):
  File "/Users/kafka/fun/tinygrad/examples/stable_diffusion.py", line 633, in <module>
    context = model.cond_stage_model.transformer.text_model(phrase).realize()
  File "/Users/kafka/fun/tinygrad/examples/stable_diffusion.py", line 463, in __call__
    x = self.encoder(x, Tensor(causal_attention_mask, device=x.device))
  File "/Users/kafka/fun/tinygrad/examples/stable_diffusion.py", line 435, in __call__
    hidden_states = l(hidden_states, causal_attention_mask)
  File "/Users/kafka/fun/tinygrad/examples/stable_diffusion.py", line 419, in __call__
    hidden_states = self.self_attn(hidden_states, causal_attention_mask)
  File "/Users/kafka/fun/tinygrad/examples/stable_diffusion.py", line 381, in __call__
    bsz, tgt_len, embed_dim = hidden_states.shape
ValueError: too many values to unpack (expected 3)
```

geohot commented:
This broke stable diffusion!

`GPU=1 python3 examples/stable_diffusion.py`
